File: yogi-cpp/yogi/path.cpp
#include "path.hpp"

using namespace std::string_literals;


namespace yogi {
namespace internal {

void throw_if_path_invalid(const std::string& str)
{
    if (str.find("//") != std::string::npos) {
        throw BadPath(str);
    }
}

} // namespace

BadPath::BadPath(const std::string& path)
: std::runtime_error("Invalid path: '"s + path + "'")
{
}

Path::Path()
{
}

Path::Path(std::string str)
: m_str(str)
{
    internal::throw_if_path_invalid(m_str);

    if (m_str.size() > 1 && m_str.back() == '/') {
        m_str.pop_back();
    }
}

const std::string& Path::to_string() const
{
    return m_str;
}

bool Path::empty() const
{
    return m_str.empty();
}

std::size_t Path::size() const
{
    return m_str.size();
}

void Path::clear()
{
    m_str.clear();
}

bool Path::is_absolute() const
{
    return !empty() && m_str.front() == '/';
}

bool Path::is_root() const
{
    return m_str == "/";
}

bool Path::operator== (const Path& rhs) const
{
    return m_str == rhs.m_str;
}

bool Path::operator!= (const Path& rhs) const
{
    return !(*this == rhs);
}

Path Path::operator/ (const Path& other) const
{
    if (other.is_absolute()) {
        throw BadPath(other.m_str);
    }

    if (is_root()) {
        return Path(m_str + other.m_str);
    }
    else {
        return Path(m_str + '/' + other.m_str);
    }
}

Path Path::operator/ (const char* other) const
{
    return *this / Path(other);
}

Path& Path::operator/= (const Path& other)
{
    *this = *this / other;
    return *this;
}

Path& Path::operator/= (const char* other)
{
    return *this /= Path(other);
}

} // namespace yogi

std::ostream& operator<< (std::ostream& os, const yogi::Path& path)
{
    return os << path.to_string();
}

File: yogi-python/yogi/private/path.py
from typing import *
import posixpath


class BadPath(Exception):
    def __init__(self, path: str):
        self._path = path

    def __str__(self):
        return 'Invalid path: \'{}\''.format(self._path)


class Path:
    def __init__(self, path: Optional[str] = None):
        self._path = path if path else ''
        if '//' in self._path:
            raise BadPath(self._path)

    def __str__(self):
        return self._path

    def __len__(self):
        return len(self._path)

    def __eq__(self, other):
        return self._path == str(other)

    def __ne__(self, other):
        return not (self == other)

    def __truediv__(self, other):
        if Path(str(other)).is_absolute:
            raise BadPath(str(other))
        return Path(posixpath.join(self._path, str(other)))

    def clear(self) -> None:
        self._path = ''

    @property
    def is_absolute(self) -> bool:
        return not len(self._path) == 0 and self._path[0] == '/'

    @property
    def is_root(self) -> bool:
        return self._path == '/'
